nget.1
.TH nget 1 "9 Aug 2000"
.SH NAME
nget \- retrieve files from NNTP hosts
.SH SYNOPSIS
.B nget [...]
.SH DESCRIPTION
.B nget
retrieves messages matching a regular expression, and decodes any files
contained within.  Multipart messages are automatically pieced together.
.SH OPTIONS
.PP
.IP "-q/--quiet"
Disable printing of merely informative messages.  Errors will still be printed.
.IP "-h/--host host"
NNTP host to use for subsequent -g commands.  Must be configured in .ngetrc.
Can reset to standard auto-choosing method with -h ""
.IP "-g/--group group"
Newsgroup to use for subsequent commands.  If a host has been specified before
with -h, it will retrieve headers only from that host.  Otherwise it will
retrieve headers for all hosts above _glevel (see configuration section for
more info on priorities.)
.IP "-G/--quickgroup group"
Like --group, but does not retrieve new headers.
.IP "-F/--flushserver host"
Flush all headers for server from current group.
.IP "-r/--retrieve regex"
Matches regex against subjects of previously selected group, and retrieves ones
that match.
.IP "-R/--expretrieve expression"
Like -r, but matches expression instead of merely a regexp.  Expression is a
postfix expression that can contain the keywords
subject, author, lines, bytes, have, req, date, anum.  Note that the --limit
argument does not affect the option, if you want to limit based on number
of lines, add it as part of the expression.
.IP "-@/--list LISTFILE"
Specify a file to load a list of command line args from.  Looks in ~/.nget3/lists/ dir by default.
.IP "-p/--path DIRECTORY"
Path to store subsequent retrieves.  Relative to path which nget was started in.
(Except in the case of inside a -@, which will be relative to the cwd at the
time of the -@.)
.IP "-T/--testmode"
Causes --retrieve to merely print out all matching files.
.IP "-t/--tries int"
Set maximum number of retries.  -1 will retry indefinatly (probably not a good
idea).
.IP "-l/--limit int"
Set the minimum number of lines a message (or total number of lines for a
multi-part message) must have to be considered for retrieval.
.IP "-s/--delay int"
Set the number of seconds to wait between retry attempts.
.IP "-i/--incomplete"
Retrieve files with missing parts.
.IP "-I/--complete"
Retrieve only files with all parts.
.IP "-k/--keep"
Keep temp files.
.IP "-K/--keep2"
Keep temp files, and don't try to decode them.
.IP "-c/--case"
Match case sensitively.
.IP "-C/--nocase"
Match case insensitively.
.IP "-d/--dupecheck FLAGS"
Check to make sure you don't already have files.  This is done in two ways.
The first ("f") is by compiling a list of all files in the current directory, then
checking against all messages to be retrieved to see if one of the filenames
shows up in the subject.  This works reasonably well, though sometimes the
filename isn't in the subject.  It can also cause problems if you happen to
have files in the directory named silly things like "a", in which case all
messages with the word "a" in them will be skipped.  However, it is still smart
enough not to skip messages that merely have a word containing "a".  The second
method ("i") is by setting a flag in the header cache that will prevent it from being
retrieved again.  You can use combos such as -dfi to check both, -dFi to only
check the flag, -dfI to only check files, etc.
.IP "-D/--nodupecheck"
Don't check either of the --dupecheck methods, retrieve any messages that
match.
.IP "-S/--tempshortnames"
Use 8.3 tempfile names (for old dos partitions, etc).
.IP "-L/--templongnames"
Use 17.3 tempfile names. 
.IP "-P/--temppath DIRECTORY"
Store temporary files in path instead of the current dir.
.IP "-N/--noconnect"
Do not connect to any server.  Useful for trying to decode as much as you have.
(if you got stuff with -K or ngetlite.)
.IP "-w/--writelite LITEFILE"
Write a list of parts to retrieve with ngetlite.
.IP "-?/--help"
Show help.
.SH CONFIGURATION
Upon startup, nget will read ~/.nget3/.ngetrc for default configuration values and host/group
aliases.  An example .ngetrc should have been included with nget.
.P
Options are specified one per line in the form:
.br
.B key=value
.br
Subsections and specified in the form 
.br
.B {section_name
.br
.B data
.br
.B }
.br
where data is any number of options.
.P
Global options that can be specified are: 
.br
Integral: limit, tries, debug, delay
.br
Float: curservmult
.br
Boolean(0/1): case, complete, dupeidcheck, dupefilecheck, quiet, tempshortnames
.P
Host configuration is done in the halias section, with a subsection for each
host containing its address, unique id, and optionally user/pass/fullxover
items.  The "id" field must be unique, and should not be changed after you have
used it (since it could leave leftovers in your cache files, unless you delete
them first).  The "id" must be greater than 0 and less than ULONG_MAX. (usually
4294967295).  In other words, don't use id 0.
.br
For example, the following host section defines a single host "host1", with
nntp authentication for user "bob", password "something", and the fullxover
option enabled. (fullxover retrieves all headers from the server, including
gaps in the header list, while normally it only retrieves headers more recent
than the latest.)
.br
.B {halias
.br
.B  {host1
.br
.B   addr=news.host1.com
.br
.B   id=3838
.br
.B   user=bob
.br
.B   pass=something
.br
.B   fullxover=1
.br
.B  }
.br
.B }
.P
Multiserver priorities are defined in the hpriority section.  Multiple priority
groups can be made, and different newsgroups can be configured to use their own
priority grouping, or they will default to the "default" group.  The hpriority
section contains a subsection for each priority group, with data items of
server=prio-multiplier, and the special items _level=float and _glevel=float.
_level sets the priority level assigned to any host not listed in the group,
and _glevel sets the required priority needed for -g to automatically use that
host.  Both _level and _glevel default to 1.0 if not specified.
The priority group "trustsizes" also has special meaning, and is used to choose which servers
reporting of article line/byte counts to trust when reporting to the user.
When retrieving articles, nget will multiply the current servers priority by
the global setting "curservmult" (default 2.0), this can be used to avoid
excesive server switching.  (set it to 1.0 if you want to disable it).
.br
For example, the following section defines the default priority group and the
trustsizes priority group.  If all hosts have a certain article, goodhost will
be most likely to be chosen, and badhost least likely.  It also sets the
default proirity level to 1.01, meaning any hosts not listed in this group will have a
priority of 1.01.  When using -g without first specifying a host, only those
with prios 1.2 or above will be selected.
.br
.B {hpriority
.br
.B  {default
.br
.B   _level=1.01
.br
.B   _glevel=1.2
.br
.B   host1=1.9
.br
.B   goodhost=2.0
.br
.B   badhost=0.9
.br
.B  }
.br
.B  {trustsizes
.br
.B   goodhost=5.0
.br
.B   badhost=0.1
.br
.B  }
.br
.B }
.P
Newsgroup aliases are defined in the galias section.  An alias can be a simple
alias=fullname data item, or  a subsection containing group= and prio= items.
.br
For example, the following galias section defines an alias of "abpl" for the
group "alt.binaries.pictures.linux", and "chocobo" for the group "alt.chocobo".
In addition, the chocobo group is assigned to use the chocoprios priority
grouping when deciding what server to retrieve from.
.br
.B {galias
.br
.B  abpl=alt.binaries.pictures.linux
.br
.B  {chocobo
.br
.B   group=alt.chocobo
.br
.B   prio=chocoprios
.br
.B  }
.br
.B }
.SH EXAMPLES

Changelog
nget (0.13.1) unstable; urgency=low

  * Use write to tempfile then rename for saving cache and midinfo.
  * Made some forgotten printfs ("was gonna add...") into PDEBUGs.
  * Use $(MAKE) instead of make for making subdirs. (Howard Jones)

 -- Matt Mueller <donut@azstarnet.com>  Mon,  8 Jan 2001 16:36:31 -0700

nget (0.13) unstable; urgency=low

  * Locks midinfo files better.  Previous versions would not have corruption,
    but would lose one of the updates, if two nget instances tried to
    update the midinfo file at the same time.
  * Added -m option to optionally make directories that do not exist for
    -p and -P.

 -- Matt Mueller <donut@azstarnet.com>  Tue,  5 Sep 2000 21:41:16 -0700
  
nget (0.12) unstable; urgency=low

  * Fixed segfault on invalid param to -F
  * Deletes tempfile on drive full error.  (Otherwise the useless
    truncated file may cause annoyances later)
  * On unequal line count errors, will now try other servers or give up, 
    instead of looping on the same one.  (See also: new linelenience option)
  * Fixed configure --with-pcre option being ignored if a path was not
    specified.
  * On uulib warning messages, cause an error and keep the temp files around
    so that the user can examine them and see what (if anything) went wrong.
  * Header cache and midinfo files now have ,cache(.gz) and ,midinfo(.gz)
    extensions, in order to prevent possible (but unlikely) problems with
    newsgroups that might have had _midinfo in the name.  More importantly,
    it allows for easy exclusion of cache files from backups.
    If you are upgrading from a previous version, run the fixnames.sh script
    to update your cache file names.
  * Added per-server linelenience option to control generation of unequal 
    line count errors.
  * Changed -q behavior, a single -q will merely disable the auto-updating
    output, to make redirecting the output (eg, logging) acceptable.
    Specifying -q twice will use the old behavior of being very quiet.

 -- Matt Mueller <donut@azstarnet.com>  Wed, 18 Aug 2000 21:47:34 -0700
  
nget (0.11) unstable; urgency=low

  * Fixed segfault on reconnecting after exception in ngetlite.
  * Added -F command to flush headers from a group.
  * Fixed const char ** compilation problem in nget.cc (Only showed up with
    newer versions of popt)

 -- Matt Mueller <donut@azstarnet.com>  Wed,  9 Aug 2000 06:33:25 -0700

nget (0.10) unstable; urgency=low

  * Fixed handling of invalid args when using libpopt.
  * Added _level and _glevel prio options, and made -g select hosts
    automatically.
  * Fixed dual \ problem in regexes. (Now you only need dual \'s for quote
    chars, rather than all chars. (once for the shell, once for nget))
  * Fixed header retrieval for new groups with fullxover=1. (only failed on 
    some servers)
  * Better error handling when selecting invalid host.
  * Better error handling for badly configured sections in .ngetrc.
  * Fixed nrange.cc compile problem in without-zlib build.

 -- Matt Mueller <donut@azstarnet.com>  Tue, 18 Apr 2000 15:55:53 -0700

nget (0.9) unstable; urgency=low

  * fixed command line arg problems when built with getopt rather that libpopt.
  * fixed creating _midinfo files with no permissions problem (caused error
    after trying to use the group for the second time on some systems).
  * fixed implicit declaration of function close() in mylockfile.h on some
    systems.
  * fixed multiple uses of -p being relative to each other rather than the
    start directory.
  * added curservmult config option (was hardcoded to 2.0 in v0.8)
  * fixed a few documentation ommisions

 -- Matt Mueller <donut@azstarnet.com>  Thu, 13 Apr 2000 22:49:22 -0700

nget (0.8) unstable; urgency=low

  * nget version 0.8 released on 2000-04-12. (sorry about that wait.)

nget.cc
/*
    nget - command line nntp client
    Copyright (C) 1999-2000  Matthew Mueller <donut@azstarnet.com>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/


#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include <signal.h>
#include <unistd.h>
#ifdef HAVE_LIBPOPT
extern "C" {
#include <popt.h>
}
#else
#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif
#endif
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <time.h>
#include "misc.h"
#include "sockstuff.h"
#include "prot_nntp.h"
#include "log.h"
#include "nget.h"
#include "datfile.h"
#include "myregex.h"

time_t lasttime;

#define NUM_OPTIONS 27
#ifndef HAVE_LIBPOPT

#ifndef HAVE_GETOPT_LONG
struct option {
	const char *name;
	int has_arg;
	int *flag;
	int val;
};
#endif

static struct option long_options[NUM_OPTIONS+1];
#define OPTIONS "-qh:g:G:r:R:@:Tt:s:l:iIcCdD?"

#else //!HAVE_LIBPOPT

static struct poptOption optionsTable[NUM_OPTIONS+1];
#endif //HAVE_LIBPOPT

struct opt_help {
	int namelen;
	char *arg;
	char *desc;
};
static opt_help ohelp[NUM_OPTIONS+1];
static int olongestlen=0;

static void addoption(char *longo,int needarg,char shorto,char *adesc,char *desc){
	static int cur=0;
	assert(cur<NUM_OPTIONS);
#ifdef HAVE_LIBPOPT
	optionsTable[cur].longName=longo;
	optionsTable[cur].shortName=shorto;
	optionsTable[cur].argInfo=needarg?POPT_ARG_STRING:POPT_ARG_NONE;
	optionsTable[cur].val=shorto;
	optionsTable[cur].arg=NULL;
#else //HAVE_LIBPOPT
	long_options[cur].name=longo;
	long_options[cur].has_arg=needarg;
	long_options[cur].flag=0;
	long_options[cur].val=shorto;
#endif //!HAVE_LIBPOPT
	ohelp[cur].namelen=longo?strlen(longo):0;
	ohelp[cur].arg=adesc;
	ohelp[cur].desc=desc;
	int l=(adesc?strlen(adesc):0)+ohelp[cur].namelen+2;
	if (l>olongestlen)
		olongestlen=l;
	cur++;
}
static void addoptions(void)
{
	addoption("quiet",0,'q',0,"supress extra info");
	addoption("host",1,'h',"HOSTALIAS","nntp host to use (must be configured in .ngetrc)");
	addoption("group",1,'g',"GROUPNAME","newsgroup");
	addoption("quickgroup",1,'G',"GROUPNAME","use group without checking for new headers");
	addoption("expretrieve",1,'R',"EXPRESSION","retrieve files matching expression(see man page)");
	addoption("retrieve",1,'r',"REGEX","retrieve files matching regex");
	addoption("list",1,'@',"LISTFILE","read commands from listfile");
	addoption("path",1,'p',"DIRECTORY","path to store subsequent retrieves");
//	addoption("mark",1,'m',"MARKNAME","name of high water mark to test files against");
//	addoption("testretrieve",1,'R',"REGEX","test what would have been retrieved");
	addoption("testmode",0,'T',0,"test what would have been retrieved");
	addoption("tries",1,'t',"INT","set max retries (-1 unlimits, default 20)");
	addoption("delay",1,'s',"INT","seconds to wait between retry attempts(default 1)");
	addoption("limit",1,'l',"INT","min # of lines a 'file' must have(default 3)");
	addoption("incomplete",0,'i',0,"retrieve files with missing parts");
	addoption("complete",0,'I',0,"retrieve only files with all parts(default)");
	addoption("keep",0,'k',0,"keep temp files");
	addoption("keep2",0,'K',0,"keep temp files and don't even try to decode them");
	addoption("case",0,'c',0,"match casesensitively");
	addoption("nocase",0,'C',0,"match incasesensitively(default)");
	addoption("dupecheck",1,'d',"FLAGS","check to make sure you haven't already downloaded files(default -dfi)");
	addoption("nodupecheck",0,'D',0,"don't check if you already have files(shortcut for -dFI)");
	addoption("tempshortnames",0,'S',0,"use 8.3 names for tempfiles");
	addoption("templongnames",0,'L',0,"use long names for tempfiles");
	addoption("temppath",1,'P',"DIRECTORY","path to store tempfiles");
	addoption("writelite",1,'w',"LITEFILE","write out a ngetlite list file");
	addoption("noconnect",0,'N',0,"don't connect, only try to decode what we have");
	addoption("help",0,'?',0,"this help");
	addoption(NULL,0,0,NULL,NULL);
};
static void print_help(void){
      printf("nget v0.8 - nntp command line fetcher\n");
      printf("Copyright 1999-2000 Matt Mueller <donut@azstarnet.com>\n");
      printf("\n\
This program is free software; you can redistribute it and/or modify\n\
it under the terms of the GNU General Public License as published by\n\
the Free Software Foundation; either version 2 of the License, or\n\
(at your option) any later version.\n\n\
This program is distributed in the hope that it will be useful,\n\
but WITHOUT ANY WARRANTY; without even the implied warranty of\n\
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n\
GNU General Public License for more details.\n\n\
You should have received a copy of the GNU General Public License\n\
along with this program; if not, write to the Free Software\n\
Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.\n");
//      printf("\nUSAGE: nget [-q] [-h host] [-g group] [-r subject regex] ?\n\n"
//	       "-q	quiet mode.  Use twice for total silence\n"
//);
	printf("\nUSAGE: nget -g group [-r file [-r file] [-g group [-r ...]]]\n");
  //this is kinda ugly, but older versions of popt don't have the poptPrintHelp stuff, and it seemed to print out a bit of garbage for me anyway...
	  for (int i=0;
#ifdef HAVE_LIBPOPT
			  optionsTable[i].longName;
#else
			  long_options[i].name;
#endif
			  i++){
			  printf("-%c  ",
#ifdef HAVE_LIBPOPT
					  optionsTable[i].shortName
#else
					  long_options[i].val
#endif
					  );
		  if (
#ifdef HAVE_LIBPOPT
				  optionsTable[i].argInfo!=POPT_ARG_NONE
#else
				  long_options[i].has_arg
#endif
				  )
			  printf("--%s=%-*s",
#ifdef HAVE_LIBPOPT
					  optionsTable[i].longName,
#else
					  long_options[i].name,
#endif
					  olongestlen-ohelp[i].namelen-1,ohelp[i].arg);
		  else
			  printf("--%-*s",
					  olongestlen,
#ifdef HAVE_LIBPOPT
					  optionsTable[i].longName
#else
					  long_options[i].name
#endif
					  );
			  printf("  %s\n",ohelp[i].desc);
	  }
#ifndef HAVE_LIBPOPT
#ifndef HAVE_GETOPT_LONG
	  printf("Note: long options not supported by this compile\n");
#endif
#endif //!HAVE_LIBPOPT
	  //cache_dbginfo();
//	  getchar();
}


c_prot_nntp nntp;

static void term_handler(int s){
	printf("\nterm_handler: signal %i, shutting down.\n",s);
	nntp.cleanup();
	exit(0);
}


string nghome;

//c_server_list servers;
//c_group_info_list groups;
//c_server_priority_grouping_list priogroups;

c_nget_config nconfig;

struct nget_options {
	int maxretry,retrydelay;
	ulong linelimit;
	int gflags,testmode,badskip,qstatus;
	c_group_info::ptr group;//,*host;
//	c_data_section *host;
	c_server *host;
//	char *user,*pass;//,*path;
	string path;
	string temppath;
	string writelite;
	//nget_options(void){path=NULL;}
	nget_options(void){}
	nget_options(nget_options &o):maxretry(o.maxretry),retrydelay(o.retrydelay),linelimit(o.linelimit),gflags(o.gflags),testmode(o.testmode),badskip(o.badskip),qstatus(o.qstatus),group(o.group),host(o.host)/*,user(o.user),pass(o.pass)*/,path(o.path),temppath(o.temppath),writelite(o.writelite){
/*		if (o.path){
			path=new char[strlen(o.path)+1];
			strcpy(path,o.path);
		}else
			path=NULL;*/
		//printf("copy path=%s(%p-%p)\n",path,this,path);
	}
//	void del_path(void){if (path){/*printf("deleteing %s(%p-%p)\n",path,this,path);*/free(path);path=NULL;}}
	//void get_path(void){/*printf("get_path\n");*/del_path();goodgetcwd(&path);}
	void do_get_path(string &s){char *p;goodgetcwd(&p);s=p;free(p);}
	void get_path(void){do_get_path(path);}
	void get_temppath(void){
		do_get_path(temppath);
		if (temppath[temppath.size()-1]!='/')
			temppath.append("/");
	}
	void parse_dupe_flags(const char *opt){
		while(*opt){
			switch (*opt){
				case 'i':gflags&= ~GETFILES_NODUPEIDCHECK;break;
				case 'I':gflags|= GETFILES_NODUPEIDCHECK;break;
				case 'f':gflags&= ~GETFILES_NODUPEFILECHECK;break;
				case 'F':gflags|= GETFILES_NODUPEFILECHECK;break;
				default:
						 throw new c_error(EX_U_FATAL,"unknown dupe flag %c",*opt);
			}
			opt++;
		}
	}
//	~nget_options(){del_path();}
};
struct s_arglist {
	int argc;
	char **argv;
};
static int do_args(int argc,char **argv,nget_options options,int sub){
	int c=0;
#ifdef HAVE_LIBPOPT
	poptContext optCon;
#else
	int opt_idx;
#endif
	int redo=0,redone=0;
	char * loptarg=NULL;
	//printf("limit:%i tries:%i case:%i complete:%i dupcheck:%i\n",linelimit,maxretry,gflags&GETFILES_CASESENSITIVE,!(gflags&GETFILES_GETINCOMPLETE),!(gflags&GETFILES_NODUPECHECK));

#ifdef HAVE_LIBPOPT
	optCon = poptGetContext(NULL, argc, argv, optionsTable, sub?POPT_CONTEXT_KEEP_FIRST:0);
#endif
	while (1){
		if (redo){
			redo=0;
		}else {
			if ((c=
#ifdef HAVE_LIBPOPT
						poptGetNextOpt(optCon)
#else //HAVE_LIBPOPT
#ifdef HAVE_GETOPT_LONG
						getopt_long(argc,argv,OPTIONS,long_options,&opt_idx)
#else
						getopt(argc,argv,OPTIONS)
#endif
#endif //!HAVE_LIBPOPT
				)<0)
				c=-12345;
#ifdef HAVE_LIBPOPT
			loptarg=poptGetOptArg(optCon);
#else
			loptarg=optarg;
#endif //HAVE_LIBPOPT
			//					break;
			redone=0;
		}
//		printf("arg:%c(%i)=%s(%p)\n",isprint(c)?c:'.',c,loptarg,loptarg);
		try {
			switch (c){
				//					case 'R':
				//						if(!group)
				//							perror("no group specified\n");
				//						nntp.nntp_queueretrieve(loptarg,linelimit,0,gflags);
				//						break;
				case 'T':
					options.testmode=1;
					printf("testmode now %i\n",options.testmode);
					break;
				//case 1:
				case 'R':
					if (!options.badskip){
						if(options.group.isnull())
							printf("no group specified\n");
						else{
							generic_pred *p=make_pred(loptarg);
							if (p){
								nntp.filec=nntp.gcache->getfiles(nntp.filec,nntp.midinfo,p,options.gflags);
								delete p;
								options.qstatus=1;
							}
						}
					}
					break;
				case 'r':
					if (!options.badskip){
						if(options.group.isnull())
							printf("no group specified\n");
						else{
							/*c_regex_nosub *reg=new c_regex_nosub(loptarg,REG_EXTENDED + ((gflags&GETFILES_CASESENSITIVE)?0:REG_ICASE));
							  if (!reg)
							  throw new c_error(EX_A_FATAL,"couldn't allocate regex");
							  if (reg->geterror()){
							  char buf[256];
							  int e=reg->geterror();
							  reg->strerror(buf,256);
							  delete reg;
							//throw new c_error(EX_A_FATAL,"regex error %i:%s",e,buf);
							printf("regex error %i:%s\n",e,buf);break;
							}
							nntp_pred *p=ecompose2(new e_land<bool>,
							new e_binder2nd<e_nntp_file_lines<e_ge<ulong> > >(linelimit),
							new e_binder2nd_p<e_nntp_file_subject<e_eq<string,c_regex_nosub*> > >(reg));

							nntp.filec=nntp.gcache->getfiles(nntp.filec,nntp.grange,p,gflags);
							delete p;
							if (!nntp.filec)
							printf("nntp_retrieve: no match for %s\n",loptarg);
							//									throw new c_error(EX_U_WARN,"nntp_retrieve: no match for %s",match);
							//								nntp.nntp_queueretrieve(loptarg,linelimit,gflags);
							qstatus=1;*/
							char *s;
							asprintf(&s,"lines %lu >= subject \"%s\" =~ &&",options.linelimit,loptarg);
							generic_pred *p=make_pred(s);
							free(s);
							if (p){
								nntp.filec=nntp.gcache->getfiles(nntp.filec,nntp.midinfo,p,options.gflags);
								delete p;
								options.qstatus=1;
							}
						}
					}
					break;
				case 'i':
					options.gflags|= GETFILES_GETINCOMPLETE;
					break;
				case 'I':
					options.gflags&= ~GETFILES_GETINCOMPLETE;
					break;
				case 'k':
					options.gflags|= GETFILES_KEEPTEMP;
					break;
				case 'K':
					options.gflags|= GETFILES_NODECODE;
					break;
				case 'c':
					options.gflags|= GETFILES_CASESENSITIVE;
					break;
				case 'C':
					options.gflags&= ~GETFILES_CASESENSITIVE;
					break;
				case 'd':
					options.parse_dupe_flags(loptarg);
					break;
				case 'D':
					options.gflags|= GETFILES_NODUPEFILECHECK | GETFILES_NODUPEIDCHECK;
					break;
				case 'S':
					options.gflags|= GETFILES_TEMPSHORTNAMES;
					break;
				case 'L':
					options.gflags&= ~GETFILES_TEMPSHORTNAMES;
					break;
				case 'N':
					options.gflags|= GETFILES_NOCONNECT;
					break;
				case 'q':
					quiet++;
					break;
				case 's':
					options.retrydelay=atoi(loptarg);
					if (options.retrydelay<0)
						options.retrydelay=1;
					printf("retry delay set to %i\n",options.retrydelay);
					break;
				case 't':
					options.maxretry=atoi(loptarg);
					if (options.maxretry==-1)
						options.maxretry=INT_MAX-1;
					printf("max retries set to %i\n",options.maxretry);
					break;
				case 'l':
					options.linelimit=atoi(loptarg);
					if (options.linelimit<0)
						options.linelimit=0;
					printf("minimum line limit set to %lu\n",options.linelimit);
					break;
				case 'w':
					options.writelite=loptarg;
					printf("writelite to %s\n",options.writelite.c_str());
					break;
				case 'P':
					if (!chdir(loptarg)){
						options.get_temppath();
						chdir(options.path.c_str());
						printf("temppath:%s\n",options.temppath.c_str());
					}else{
						printf("could not change temppath to %s\n",loptarg);
						return -1;
					}
					break;
				case 'p':
					if (!chdir(loptarg)){
						options.get_path();
						printf("path:%s\n",options.path.c_str());
					}else{
						printf("could not change to %s\n",loptarg);
						return -1;
					}
					break;
				case ':':
				case '?':
					print_help();
					return 1;
				default:
					if (options.qstatus){
						if (!options.badskip) nntp.nntp_retrieve(!options.testmode,options.gflags,options.temppath,options.writelite);
						options.qstatus=0;
					}
					switch (c){
						case '@':
#ifdef HAVE_LIBPOPT
							{
								const char *paths[2]={(nghome + "lists/").c_str(),NULL};
								const char *filename=fsearchpath(loptarg,paths,FSEARCHPATH_ALLOWDIRS);
								c_file_fd f;
								f.initrbuf(2048);
								if (filename && !f.open(filename,O_RDONLY)){
									int totargc=0;
									int pr;
									s_arglist larg;
									list<s_arglist> arglist;
									while (f.bgets()>0){
										if(!(pr=poptParseArgvString(f.rbufp(),&larg.argc,&larg.argv))){
											arglist.push_back(larg);
											totargc+=larg.argc;
										}else printf("poptParseArgvString:%i\n",pr);
									}
									f.close();
									if (!arglist.empty()){
										int tc=0,ic=0;
										s_arglist targ;
										larg.argc=totargc;
										larg.argv=(char**)malloc((totargc+1)*sizeof(char**));
										list<s_arglist>::iterator it=arglist.begin();
										for (;it!=arglist.end();++it){
											targ=*it;
											for(ic=0;ic<targ.argc;ic++,tc++)
												larg.argv[tc]=targ.argv[ic];
										}

										do_args(larg.argc,larg.argv,options,1);
										if (options.host){//####here we reset the stuff that may have been screwed in our recursiveness.  Perhaps it should reset it before returning, or something.. but I guess this'll do for now, since its the only place its called recursively.
											nntp.nntp_open(options.host);
											if (!options.group.isnull())
												nntp.nntp_group(options.group,0);
										}
										if (!chdir(options.path.c_str())){
											printf("path:%s\n",options.path.c_str());
										}else{
											printf("could not change to %s\n",options.path.c_str());
											return -1;
										}
										for (it=arglist.begin();it!=arglist.end();++it)
											free((*it).argv);
										free(larg.argv);
									}
								}else{
									printf("error opening %s: %s(%i)\n",filename?:loptarg,strerror(errno),errno);errno=0;
								}
								if (filename && filename!=loptarg)
									free(const_cast<char*>(filename));//const_cast away the warning
							}
#else
							printf("This option is only available when libpopt is used.\n");
#endif
							break;
						case 'g':
							if (options.badskip<2){
								if (options.host==NULL){
									printf("null host, trying %s\n",getenv("NNTPSERVER"));
									if (getenv("NNTPSERVER")){
										options.host=nconfig.getserver(getenv("NNTPSERVER"));//###### TODO: KLUDGE since host prio stuff isn't finished.
										nntp.nntp_open(options.host);
									}
								}
								options.group=nconfig.getgroup(loptarg);
//								if (!galias || !(options.group=galias->getitema(loptarg)))
//									options.group=loptarg;
								nntp.nntp_group(options.group,1);
								if (options.badskip)
									options.badskip=0;
							}
							break;
						case 'G':
							if (options.badskip<2){
								options.group=nconfig.getgroup(loptarg);
//								if (!galias || !(options.group=galias->getitema(loptarg)))
//									options.group=loptarg;
								nntp.nntp_group(options.group,0);
								if (options.badskip)
									options.badskip=0;
							}
							break;
						case 'h':{
									 /*options.user=loptarg,options.pass=NULL;
									 goodstrtok(&options.user,' ');
									 if (options.user){
										 options.pass=options.user;
										 goodstrtok(&options.pass,' ');
									 }
									 options.badskip=0;
									 if (!halias || !(options.host=halias->getitema(loptarg)))
										 options.host=loptarg;
									 nntp.nntp_open(options.host,options.user,options.pass);*/
									 options.host=nconfig.getserver(loptarg);
//									 options.host=halias->getsection(loptarg);
									 nntp.nntp_open(options.host);
								 }
								 break;
						case -12345:
								 //								if (!badskip){
								 //									nntp.nntp_retrieve(!testmode);
								 //								}
								 return 0;
						default:
								 print_help();
								 return 1;
					}
			}
		}catch(c_error *e){
			int n=e->num;
			printf("caught exception %i: %s",n,e->str);
			delete e;
			if (n<EX_FATALS){
				if(redone<options.maxretry){
					redo=1;redone++;
					printf(" (trying again. %i)\n",redone);
					if (options.retrydelay)
						sleep(options.retrydelay);
				}else{
					printf("\n");
					redo=0;redone=0;
				}
			}else{
				if (n==EX_A_FATAL){
					printf(" (fatal application error, exiting..)\n");
					exit(-1);
				}else{
					printf(" (fatal, aborting..)\n");
					if (n==EX_T_FATAL)
						options.badskip=2;
					//else if (n==EX_U_FATAL)
					else
						options.badskip=1;
				}
			}
		}
	}
	//		printf("quiet=%i host=%s\n",quiet,nntp_server);
	//		for (int i=optind;i<argc;i++)
	//			switch (mode){
	//				default:
	//					printf("%s\n",argv[i]);
	//					nntp_get(nntp_server,group,argv[i]);
	//			}

#ifdef HAVE_LIBPOPT
	poptFreeContext(optCon);
#endif
}
int main(int argc,char ** argv){
	try {
		//	atexit(cache_dbginfo);
		//string hi="aoeu";
		//	char *aoeu=new char[5];
		init_my_timezone();
		//	nntp.nntp_open(getenv("NNTPSERVER")?:"localhost",NULL,NULL);
		addoptions();
		signal(SIGTERM,term_handler);
		signal(SIGHUP,term_handler);
		signal(SIGINT,term_handler);
		signal(SIGQUIT,term_handler);
		{
			char *home;
			home=getenv("NGETHOME");
			if (home)
				nghome=home;
			else {
				home=getenv("HOME");
				if (!home){
					printf("HOME environment var not set.");
					exit(1);
				}
				nghome=home;
				nghome.append("/.nget3/");
			}
		}


		time(&lasttime);
		srand(lasttime);
		if (argc<2){
			print_help();
		}
		else {
			nget_options options;
			options.maxretry=20;
			options.retrydelay=1;
			options.badskip=0;
			options.linelimit=3;
			options.gflags=0;
			options.testmode=0;
			options.qstatus=0;
			options.group=NULL;
			options.host=NULL;
			//		options.user=NULL;
			//		options.pass=NULL;
			{
				c_data_file cfg;
				c_data_section *galias,*halias,*hpriority;
				cfg.setfilename((nghome + ".ngetrc").c_str());
				cfg.read();
				halias=cfg.data.getsection("halias");
				hpriority=cfg.data.getsection("hpriority");
				galias=cfg.data.getsection("galias");
				if (!halias) {
					fprintf (stderr,"no halias section (or no ngetrc at all), please configure first. (see man nget)\n");
					return 1;
				}
				nconfig.setlist(halias,hpriority,galias);
				int t;
				/*			if (!halias || !(options.host=halias->getitema("default")))
							options.host=getenv("NNTPSERVER")?:"";
							nntp.nntp_open(options.host,NULL,NULL);*/
				/*			options.host=halias->getsection("default");
							nntp.nntp_open(options.host);*/
				cfg.data.getitemi("debug",&debug);
				cfg.data.getitemi("quiet",&quiet);
				cfg.data.getitemul("limit",&options.linelimit);
				cfg.data.getitemi("tries",&options.maxretry);
				cfg.data.getitemi("delay",&options.retrydelay);
				if (!cfg.data.getitemi("case",&t) && t==1)
					options.gflags|= GETFILES_CASESENSITIVE;
				if (!cfg.data.getitemi("complete",&t) && t==0)
					options.gflags|= GETFILES_GETINCOMPLETE;
				if (!cfg.data.getitemi("dupeidcheck",&t) && t==0)
					options.gflags|= GETFILES_NODUPEIDCHECK;
				if (!cfg.data.getitemi("dupefilecheck",&t) && t==0)
					options.gflags|= GETFILES_NODUPEFILECHECK;
				if (!cfg.data.getitemi("tempshortnames",&t) && t==1)
					options.gflags|= GETFILES_TEMPSHORTNAMES;
			}
			options.get_path();
			nntp.initready();
			do_args(argc,argv,options,0);
		}
	}catch(c_error *e){
		int n=e->num;
		printf("main(): caught exception %i: %s",n,e->str);
		delete e;
	}

	return 0;
}

